test/app-luatest/trigger_module_test.lua

```diff
@@ -0,0 +1,498 @@
+    trigger.set(event, 'f1', f1)
+    trigger.call(event)
+    t.assert_equals(order, {1}, "Correct trigger invocation when 1st trigger \
+                                 clears 2nd")
+end
+
+g.test_pairs = function()
+    local trigger = require('trigger')
+
+    local state = 0
+    local function handler(inc)
+        state = state + inc
+    end
+    local function handler_method(self, inc)
+        state = state + inc
+    end
+    local a, b, c = generate_callable_objects(handler_method)
+    local handlers = {handler, c, b, a, handler, c, a, handler, b, handler}
+    local event = 'event_name'
+    local trigger_prefix = 'trigger_name.'
+
+    for i = #handlers, 1, -1 do
+        trigger.set(event, trigger_prefix..tostring(i),  handlers[i])
+    end
+
+    local idx = 0
+    local sum = 0
+    for name, h in trigger.pairs(event) do
+        idx = idx + 1
+        sum = sum + idx
+        t.assert_equals(name, trigger_prefix .. tostring(idx))
+        t.assert_equals(h, handlers[idx])
+        h(idx)
+    end
+    t.assert_equals(idx, #handlers)
+    t.assert_equals(state, sum)
+    for i = 1, #handlers do
+        trigger.del(event, trigger_prefix .. tostring(i))
+    end
+end
+
+-- Check that a trigger deleted during pairs does not break anything and
+-- isn't yielded by pairs after its deletion.
+g.test_del_during_pairs = function()
+    local trigger = require('trigger')
+
+    local event = 'event'
+    local trigger_num = 100
+    local break_point = math.random(5, 95)
+
+    t.assert_equals(trigger.pairs(event), nil)
+
+    local history = {}
+    local function handler_method(self)
+        table.insert(history, self.id)
+    end
+    local mt = {__call = handler_method}
+    for i = trigger_num, 1, -1 do
+        local a = {id = i}
+        setmetatable(a, mt)
+        trigger.set(event, tostring(i), a)
+    end
+
+    local idx = 0
+    local iter = {trigger.pairs(event)}
+    for _, handler in unpack(iter) do
+        idx = idx + 1
+        handler()
+        if idx == break_point then
+            break
+        end
+    end
+    trigger.del(event, tostring(math.random(1, break_point)))
+    trigger.del(event, tostring(break_point))
+    local deleted_id = math.random(break_point + 1, 100)
+    trigger.del(event, tostring(deleted_id))
+
+    for _, handler in unpack(iter) do
+        handler()
+    end
+
+    t.assert_equals(#history, trigger_num - 1)
+    local history_idx = 0
+    for i = 1, trigger_num do
+        if i ~= deleted_id then
+            history_idx = history_idx + 1
+            t.assert_equals(history[history_idx], i)
+            trigger.del(event, tostring(i))
+        end
+    end
+end
+
+-- Check that pairs is successfully stopped after all triggers registered
+-- on the event are dropped
+g.test_clear_event_during_pairs = function()
+    local trigger = require('trigger')
+
+    local event = 'event'
+    local trigger_num = 100
+    local break_point = math.random(5, 95)
```

Comment:
Random tests are difficult to debug because it isn't easy to figure out which seed caused a failure. Can you rewrite the tests in a deterministic way?

Reply:
100 triggers is overkill so I reduced this number to 10, but I did not removed random form the tests.